floors/rule.go

```diff
@@ -0,0 +1,556 @@
+	VideoMedia          string = "video-instream"
+	VideoOutstreamMedia string = "video-outstream"
+	AudioMedia          string = "audio"
+	NativeMedia         string = "native"
+)
+
+// getFloorCurrency returns floors currency provided in floors JSON,
+// if currency is not provided then defaults to USD
+func getFloorCurrency(floorExt *openrtb_ext.PriceFloorRules) string {
+	var floorCur string
+
+	if floorExt != nil && floorExt.Data != nil {
+		if floorExt.Data.Currency != "" {
+			floorCur = floorExt.Data.Currency
+		}
+
+		if len(floorExt.Data.ModelGroups) > 0 && floorExt.Data.ModelGroups[0].Currency != "" {
+			floorCur = floorExt.Data.ModelGroups[0].Currency
+		}
+	}
+
+	if len(floorCur) == 0 {
+		floorCur = "USD"
+	}
+
+	return floorCur
+}
+
+// getMinFloorValue returns floorMin and floorMinCur,
+// values provided in impression extension are considered over floors JSON.
+func getMinFloorValue(floorExt *openrtb_ext.PriceFloorRules, imp *openrtb_ext.ImpWrapper, conversions currency.Conversions) (float64, string, error) {
+	var err error
+	var rate float64
+	var floorCur string
+	floorMin := roundToFourDecimals(floorExt.FloorMin)
+	floorMinCur := floorExt.FloorMinCur
+
+	impFloorMin, impFloorCur, err := getFloorMinAndCurFromImp(imp)
+	if err == nil {
+		if impFloorMin > 0.0 {
+			floorMin = impFloorMin
+		}
+		if impFloorCur != "" {
+			floorMinCur = impFloorCur
+		}
+
+		floorCur = getFloorCurrency(floorExt)
+		if floorMin > 0.0 && floorMinCur != "" {
+			if floorExt.FloorMinCur != "" && impFloorCur != "" && floorExt.FloorMinCur != impFloorCur {
+				glog.Warning("FloorMinCur are different in floorExt and ImpExt")
+			}
+			if floorCur != "" && floorMinCur != floorCur {
+				rate, err = conversions.GetRate(floorMinCur, floorCur)
+				floorMin = rate * floorMin
+			}
+		}
+		floorMin = roundToFourDecimals(floorMin)
+	}
+	if err != nil {
+		return floorMin, floorCur, fmt.Errorf("Error in getting FloorMin value : '%v'", err.Error())
+	} else {
+		return floorMin, floorCur, err
+	}
+}
+
+// getFloorMinAndCurFromImp returns floorMin and floorMinCur from impression extension
+func getFloorMinAndCurFromImp(imp *openrtb_ext.ImpWrapper) (float64, string, error) {
+	var floorMin float64
+	var floorMinCur string
+
+	impExt, err := imp.GetImpExt()
+	if impExt != nil {
+		impPrebidExt := impExt.GetPrebid()
+		if impPrebidExt != nil && impPrebidExt.Floors != nil {
+			if impPrebidExt.Floors.FloorMin > float64(0) {
+				floorMin = impPrebidExt.Floors.FloorMin
+			}
+
+			if impPrebidExt.Floors.FloorMinCur != "" {
+				floorMinCur = impPrebidExt.Floors.FloorMinCur
+			}
+		}
+	}
+	return floorMin, floorMinCur, err
+}
+
+// updateImpExtWithFloorDetails updates floors related details into imp.ext.prebid.floors
+func updateImpExtWithFloorDetails(imp *openrtb_ext.ImpWrapper, matchedRule string, floorRuleVal, floorVal float64) error {
+	impExt, err := imp.GetImpExt()
+	if err != nil {
+		return err
+	}
+	extImpPrebid := impExt.GetPrebid()
```

Comment:
Nitpick: To follow the naming structure of the function name, maybe this could be renamed to be `impExtPrebid`?

Reply:
renamed to `impExtPrebid`